### ISSUE TITLE
[Feature Request] TwinFlow: Realizing One-step Generation on Large Models with Self-adversarial Flows (1-2 step distillation without any teacher/extra models, requires way less training than DMD)

### ISSUE THREAD

**Issue by MeiYi-dev** [bghira/SimpleTuner#2139]
https://zhenglin-cheng.com/twinflow/

https://arxiv.org/abs/2512.05150

https://github.com/inclusionAI/TwinFlow

https://huggingface.co/inclusionAI/TwinFlow

Compared to DMD/DCM the biggest advantage is it preserves the diversity of the model, doesn't require extra (teacher, etc) models and requires way less training compared to DMD/DCM.

**Comment by kabachuha** [comment]
@MeiYi-dev I think this may be out of scope for this project. After all, it's for tuning (LoRA, community tunes), not for creating the base models (like making a DMD). The distillations are need to be made once, and then forgotten about. Additionally, such large scale projects can be launched by companies, laboratories and wealthy individuals, not hobbyists. Let's not make the life of the maintainer harder 😬

**Comment by bghira** [comment]
SimpleTuner already has distillation primitives and ODE endpoint caching/sampling including DMD, LCM, and reflow.

**Comment by MeiYi-dev** [comment]
> [@MeiYi-dev](https://github.com/MeiYi-dev) I think this may be out of scope for this project. After all, it's for tuning (LoRA, community tunes), not for creating the base models (like making a DMD). The distillations are need to be made once, and then forgotten about. Additionally, such large scale projects can be launched by companies, laboratories and wealthy individuals, not hobbyists. Let's not make the life of the maintainer harder 😬

The only reason I requested this is because the project already supports some distiillation training. TwinFlow is simply another way to distill any model just like DMD, except it's way more usable for hobbyists, TwinFlow is much simpler to integrate than DMD/DCM and doesn't take that many training steps compared to DMD/DCM.

**Comment by kabachuha** [comment]
@MeiYi-dev got it! Thanks for the explaination

**Comment by bghira** [comment]
holy fuck 😂 this is deceptively simple looking. it requires an update to every models' forward pass to accept negative timestep conditioning information, and add a new negative time embed. they pass sigma for the main time embed and then inject a sign via extra embed, aux_time_embed when running the self-adversarial pass (for Qwen).

so basically every model that this supports would require this modification at inference *and* training time.

just a heads-up in case you were thinking this would be a seamless transition. there's quite a bit of ecosystem support that's lacking for this changeset.

**Comment by bghira** [comment]
https://github.com/bghira/SimpleTuner/pull/2159

since it requires quite a lot of training to actually test this one, i'm not really sure how to handle that 😅 

i've got the loss terms all working as far as i can tell, my loss values are in the same range as their upstream implementation.

they don't have Qwen training code just yet, only the aux embed and inference stuff so i did my best and aligned to their MNIST implementation where ambiguity came up.

all flow-matching transformer models except omnigen and lumina2 were simple enough to keep repeating the pattern, but i haven't even run through them to do smoke tests on their functionality.

**Comment by MeiYi-dev** [comment]
> [#2159](https://github.com/bghira/SimpleTuner/pull/2159)
> 
> since it requires quite a lot of training to actually test this one, i'm not really sure how to handle that 😅
> 
> i've got the loss terms all working as far as i can tell, my loss values are in the same range as their upstream implementation.
> 
> they don't have Qwen training code just yet, only the aux embed and inference stuff so i did my best and aligned to their MNIST implementation where ambiguity came up.
> 
> all flow-matching transformer models except omnigen and lumina2 were simple enough to keep repeating the pattern, but i haven't even run through them to do smoke tests on their functionality.

I am going to do tests on Kandinsky 5.0 Lite 2B within 2 weeks, once all of this works I will move on to a large scale finetune of the K5 Pro model.
